### PR TITLE
모델 트레이닝 코드 작성

### PR DESCRIPTION
**Pull request by dailyhenie**
# Pull Request
레파지토리에 기여해주셔서 감사드립니다.

해당 PR을 제출하기 전에 아래 사항이 완료되었는지 확인 부탁드립니다:
- [x] 작성한 코드가 어떤 에러나 경고없이 빌드가 되었나요?
- [x] 충분한 테스트를 수행하셨나요?

## 1. 해당 PR은 어떤 내용인가요?
* dcgan 모델 구현과 학습 코드를 담고 있습니다

## 2. PR과 관련된 이슈가 있나요?
#2 #3 #5

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    def split_dataset(self) -> [Dataset, Dataset]:
+        # split by fixed validation size
+        return random_split(
+            self.dataset,
+            [
+                self.conf.params.validation_size,
+                len(self.dataset) - self.conf.params.validation_size,
+            ],
+            generator=torch.Generator().manual_seed(42),
```
Comment (ssaru) [inline src/dataset.py]: 오... `torch.Generator()` 처음 알았습니다.
감사합니다.

그런데, 별도의 Pytorch의 난수 생성기를 사용하신 목적이 무엇이실까요?

File: src/model/net.py [hunk 2]
```diff
@@ -1,4 +1,118 @@
...
+        )
+
+    def _conv(
+        self, c_in: int, c_out: int, kernel_size: int = 5, stride: int = 2
+    ) -> nn.Sequential:
+        return nn.Sequential(
+            nn.Conv2d(c_in, c_out, kernel_size, stride, padding=2),
+            nn.BatchNorm2d(c_out),
+            nn.LeakyReLU(0.2, inplace=True),
+        )
```
Comment (ssaru) [inline src/model/net.py]: 네이밍을 `_conv_block` 형태로 하시면 어떨까요?

`_conv`라는 이름에 대해서는 저는 Convolutional NN이라고 이해했는데요.
실제 내용을 보니 Conv2d - BatchNorm - Activation이 함께 있어서
conv block이 더 적합하다는 느낌이 들었습니다.

다른 분들은 어떻게 생각하시는지 이야기를 들어보면 더 좋을 것 같네요.

File: src/model/net.py [hunk 3]
```diff
@@ -1,4 +1,118 @@
...
+    def forward(self, z: torch.Tensor) -> torch.Tensor:
+        d = self._model(z)
+        return d
+
+
+class Discriminator(nn.Module):
+    def __init__(self, hparams):
+        super(Discriminator, self).__init__()
+        self.hparams = hparams
+        self.main = nn.Sequential(
```
Comment (seopbo) [inline src/model/net.py]: 코드 깔끔하게 잘짜시네요! 다만 저는 member 이름으로 `main`은 조금 꺼려져서, 전체 큰 뭉텅이?의 operation은 만들었다는 의미로 무언가로 대체하면 좋을 것 같습니다. 예를 들면 `self.main` -> `self.ops` 그렇게 생각하는 이유는 내부적으로 사용하는 부분에 대해서는 이미 `self._conv` 이런식으로 구현을 해놓으셔서 그렇습니다. (물론 개인의견입니다. 사실 `main`보다 적합한 게 있는 지 잘 모르겟기도하네요..)

File: src/runner.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        real_images, labels = batch
+
+        # optimizer index에 따른 구현
+        # generator
+        if optimizer_idx == 0:
+            z = torch.randn(real_images.size(0), self.hparams.size_of_latent_vector)
+
+            self.fake_images = self.forward(z)
```
Comment (ssaru) [inline src/runner.py:89]: `self.fake_images`라는 변수를 사용하셨는데, 혹시 `fake_images`가 상태값으로 메모리에 저장해야하는 이유가 별도로 있을까요?

만약 꼭 상태값으로 저장하지 않아도 되는거면 `self`는 안써도 될 것 같아서요!

Comment (ssaru) [reply src/runner.py:89]: 아 밑에 optimizer index가 바뀌는 것을 고려하기 위해서 상태값을 사용하는 것이군요!

음... 조금 더 우아하게 사용할 수 없으려나요.... 뭔가 더 이쁘게 쓸 수 있을 것 같다는 느낌이.....!!!
들지만..... 생각이 나지 않네요 ㅠㅠ

File: src/runner.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            fake = torch.zeros(real_images.size(0), 1)
+            fake_loss = self.adversarial_loss(
+                self.discriminator(self.fake_images.detach()), fake
+            )
+
+            d_loss = (real_loss + fake_loss) / 2
+            tqdm_dict = {"d_loss": d_loss}
+            output = OrderedDict(
+                {"loss": d_loss, "progress_bar": tqdm_dict, "log": tqdm_dict}
+            )
```
Comment (ssaru) [inline src/runner.py]: 음... `OrderedDict`를 꼭 써야하는 이유가있으려나요?

File: tests/test_ops.py [hunk 6]
```diff
@@ -10,18 +10,3 @@ def fix_seed():
     pytorch_lightning.seed_everything(777)
     torch.backends.cudnn.deterministic = True
     torch.backends.cudnn.benchmark = False
```
Comment (ssaru) [inline tests/test_ops.py]: 혜림님 테스트 코드 짭시다! ㅋㅋㅋㅋㅋ

**Review by ssaru — COMMENTED** [review]
@hyerim1048 코멘트 완료했습니다!

File: src/model/net.py [hunk 7]
```diff
@@ -1,4 +1,118 @@
...
+    return int(math.ceil(float(size) / float(stride)))
+
+
+class Generator(nn.Module):
+    def __init__(self, hparams: dict, output_size: int = 32) -> None:
+        super(Generator, self).__init__()
+        self.hparams = hparams
+        self.output_size = output_size
+
+        self._model = self._build_model()
```
Comment (moon1ite) [inline src/model/net.py]: `hparams`나 `output_size`와 다르게 `model`은 private 으로 선언하신 이유가 있을까요?
개인적으로는 Generator 라는 class에 가장 어울리는 변수는 model 이라서 public 변수로 선언하는 것이 더 적합하다고 생각합니다.

File: src/model/net.py [hunk 8]
```diff
@@ -1,4 +1,118 @@
...
+
+    def _build_model(self) -> nn.Sequential:
+
```
Comment (moon1ite) [inline src/model/net.py:61]: style 상 빈줄 제거가 필요합니다

File: src/model/net.py [hunk 9]
```diff
@@ -1,4 +1,118 @@
...
+            nn.ConvTranspose2d(
+                self.hparams.num_gen_filters * 1,
+                self.hparams.num_channels,
+                5,
+                2,
+                padding=2,
+                output_padding=1,
+            ),
+            nn.Tanh(),  # [128, 3, 32, 32]
+        )
```
Comment (moon1ite) [inline src/model/net.py]: style 문제이긴한데요, 개인적으로 python은 "읽기 쉬움"이 이 언어의 가장 큰 목적이자 특징이라고 생각하고 있어서,
`nn.Sequential` 을 return 으로 곧장 적는 것보다, 이 것의 의미하는 바를 변수로 받아서 이를 return 하는 것이 더 깔끔하다고 생각해요!
```suggestion
        model = nn.Sequential(
            # project and reshape
            nn.Linear(
                self.hparams.size_of_latent_vector,
                self.hparams.num_gen_filters * 8 * s_16 * s_16,
                bias=False,
            ),
            Reshape(-1, self.hparams.num_gen_filters * 8, s_16, s_16),
            nn.ReLU(True),
            self._conv_transpose(
                self.hparams.num_gen_filters * 8,
                self.hparams.num_gen_filters * 4,
            ),  # [128, 256, 4, 4]
            self._conv_transpose(
                self.hparams.num_gen_filters * 4,
                self.hparams.num_gen_filters * 2,
            ),  # input[128, 128, 8, 8]
            self._conv_transpose(
                self.hparams.num_gen_filters * 2,
                self.hparams.num_gen_filters * 1,
            ),  # input[128, 64, 16, 16]
            nn.ConvTranspose2d(
                self.hparams.num_gen_filters * 1,
                self.hparams.num_channels,
                5,
                2,
                padding=2,
                output_padding=1,
            ),
            nn.Tanh(),  # [128, 3, 32, 32]
        )
        return model
```

File: src/model/net.py [hunk 10]
```diff
@@ -1,4 +1,118 @@
...
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from src.model.ops import *
+
+
+def conv_out_size(size, stride: int = 2) -> int:
+    return int(math.ceil(float(size) / float(stride)))
```
Comment (moon1ite) [inline src/model/net.py]: 이 함수는 `Generator`에서만 쓰이는 것 같네요!
여기에 있으면 `Generator`와 `Discriminator`모두에 있어야 할 것 같다는 인상을 주는데요,
혹시 `Generator` 클래스 안에 private 함수로 넣는 것은 어떨까요?

```suggestion
class Generator(nn.Module):
    ...
    def _conv_out_size(self, size, stride: int = 2) -> int:
        return int(math.ceil(float(size) / float(stride)))
```

File: src/model/net.py (hunk 2 above)
Comment (moon1ite) [reply src/model/net.py]: 저도 @ssaru 의견에 동의합니다 😃

File: src/runner.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        self.hparams = hparams
+        self.generator = generator
+        self.discriminator = discriminator
+
+    def configure_optimizers(self) -> (List, List):
+        lr = self.hparams.lr
+
+        g_optimizer = torch.optim.Adam(self.generator.parameters(), lr=lr)
+        d_optimizer = torch.optim.Adam(self.discriminator.parameters(), lr=lr)
+        return [g_optimizer, d_optimizer], []  # optimizer list 전달
```
Comment (moon1ite) [inline src/runner.py]: 여기 return 에 빈 list 까지 반환하는 것은 왜인가요? document 가 있으면 이해가 쉬울 것 같아요!

File: train.py [hunk 12]
```diff
@@ -1,10 +1,56 @@
...
+def run(conf: DictConfig) -> None:
+    train_dataloader, val_dataloader = get_dataloader(conf)
+    model_G = Generator(hparams=conf.model.params)
+    model_D = Discriminator(hparams=conf.model.params)
+    runner = Runner(conf.model.params, model_G, model_D)
+    trainer = pl.Trainer(logger=wandb_logger)
+    trainer.fit(runner, train_dataloader=train_dataloader)
+    # trainer.save_checkpoint("DCGANADam-32-0.001.pth")
+    # wandb.save("DCGANADam-32-0.001.pth")
+    # , val_dataloaders=val_dataloader
```
Comment (moon1ite) [inline train.py]: 앗, 이 부분이 주석처리되면 안될 것 같다는 생각이 드는데 @_@ ㅋㅋㅋ

**Review by moon1ite — CHANGES_REQUESTED** [review]
@hyerim1048 리뷰완료 했습니다! 한번 확인해주세요 ㅎㅎ

File: src/dataset.py (hunk 1 above)
Comment (dailyhenie) [reply src/dataset.py]: @ssaru 논문에서 validation dataset 10,000개를 먼저 training set에서 분리하는데, pytorch 에서 제공하는 train dataset에서 고정된 10000개의 validation dataset을 뽑고 싶었습니당!!

File: src/model/net.py (hunk 2 above)
Comment (dailyhenie) [reply src/model/net.py]: @ssaru , @inmoonlight  감사합니다!! 아래 커밋에서 각 레이어들의 이름을 다시 정리했어요 😊

File: src/model/net.py (hunk 3 above)
Comment (dailyhenie) [reply src/model/net.py]: @aisolab 감사합니다!! self.main ->  self.layers 로 변경했어요 !! 👀

File: src/runner.py (hunk 4 above)
Comment (dailyhenie) [reply src/runner.py:89]: 흡.. 조금 더 고민해 봐야겠어요 😢

File: src/runner.py (hunk 5 above)
Comment (dailyhenie) [reply src/runner.py]: @ssaru 꼭 써야하는 이유가 없는 듯해 제거했습니다🙈

File: tests/test_ops.py (hunk 6 above)
Comment (dailyhenie) [reply tests/test_ops.py]: 예에 매우 간단하게.. 추가했습니다 ㅋㅋㅋㅋㅋ

File: src/model/net.py (hunk 7 above)
Comment (dailyhenie) [reply src/model/net.py]: 각 generator, discriminator의 변수명, 함수명들을 제대로 정리하지 않아가지구, 다시 정리했습니다!! 
public으로 다시 선언했어요 😁

File: train.py (hunk 12 above)
Comment (dailyhenie) [reply train.py]: model 저장하는 callback 으로 대체했습니다! 감사해요 ☺️

File: src/model/net.py (hunk 9 above)
Comment (dailyhenie) [reply src/model/net.py]: 감사합니다! 저도 제안해주신 부분이 조금 더 직관적인 것 같아 변경했습니다 @inmoonlight

File: src/model/net.py (hunk 10 above)
Comment (dailyhenie) [reply src/model/net.py]: 요 부분은 크게 필요하지 않을 것 같아 제거했어요!! 감사합니다

File: src/runner.py (hunk 11 above)
Comment (dailyhenie) [reply src/runner.py]: lr_schedulers list 를 pytorch lightning 에서 받아서 그렇게 작성했습니당!! document 는 추가적으로 작성해볼게요!!

**Comment by dailyhenie** [comment]
@ssaru , @aisolab , @inmoonlight  앗 멘션 거는게 늦었지만, PR 수정 완료했습니다!! 🙏

File: src/runner.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+            # save models
+            filename = f"{self.filename_prefix}_epoch_{epoch}.ckpt"
+            ckpt_path = f"{self.file_path}/{filename}"
+            torch.save(
+                {
+                    "generator": pl_module.generator.state_dict(),
+                    "discriminator": pl_module.discriminator.state_dict(),
+                },
+                ckpt_path,
+            )
```
Comment (ssaru) [inline src/runner.py:39]: 오... 덕분에 재밌는거 배웠습니다.
이렇게도 할 수 있군요.

**Review by ssaru — APPROVED** [review]
@hyerim1048 
LGTM!
고생하셨습니다!
